cpp/src/rolling/rolling.cu

```diff
@@ -1283,22 +1287,80 @@ size_t multiplication_factor(cudf::data_type const& data_type)
   }
 }
 
+// Given a single, ungrouped timestamp column, return the indices corresponding
+// to the first null timestamp, and (one past) the last null timestamp.
+// The input column is sorted, with all null values clustered either
+// at the beginning of the column or at the end.
+// If no null values are founds, null_begin and null_end are 0.
+std::tuple<size_type, size_type> get_null_bounds_for_timestamp_column(
+  column_view const& timestamp_column)
+{
+  auto p_timestamps_device_view = column_device_view::create(timestamp_column);
+  auto timestamp_begin          = thrust::make_counting_iterator(0);
+  auto timestamp_end            = thrust::make_counting_iterator(timestamp_column.size());
+
+  // TODO: Handle case where timestamp_column is not nullable.
+
+  auto nulls_begin = timestamp_column.has_nulls()
+                       ? thrust::find_if(thrust::device,
+                                         timestamp_begin,
+                                         timestamp_end,
+                                         [timestamps = *p_timestamps_device_view] __device__(
+                                           auto i) { return timestamps.is_null_nocheck(i); })
+                       : timestamp_begin;
+  auto nulls_begin_idx = nulls_begin - timestamp_begin;
+
+  auto nulls_end = timestamp_column.has_nulls()
+                     ? thrust::find_if(thrust::device,
+                                       nulls_begin,
+                                       timestamp_end,
+                                       [timestamps = *p_timestamps_device_view] __device__(auto i) {
+                                         return !timestamps.is_null_nocheck(i);
+                                       })
+                     : timestamp_begin;
```

Comment:
Given that the input is sorted (at least in terms of null-ness) this part can be made much more efficient. You can use thrust::upper_bound to find the first or last null element (depending on whether the nulls are at the beginning or at the end). This is a single binary search, instead of the two linear searches used here.
I think something like this should work:
```
if (not nullable or (first not null and last not null return)) return {0,0};
if (first is null) return {0, upper_bound(...., []{return !timestamps.is_null_nocheck(i)} - timestamp_begin)};
else return {upper_bound(...., []{return timestamps.is_null_nocheck(i)} - timestamp_begin, timestamp_column.size())}; // last must be null
```
I left out a lot of the syntax, let me know if it's not possible to follow and I can expand the code.

Reply:
Sorry, what is the recommended way to check if a row at an index is null, from `__host__` code?

Reply:
Sorry, it took me a little effort to digest. Good one, this. Thank you, @vuule.

I think I should also be able to do something similar in the grouped case. I'm taking a crack at this right now.

Reply:
Hmm... My comment from before might still be relevant, though. Can `upper_bound()` be used on a sorted list where nulls may be clustered either `FIRST` or `LAST`, independently of `ASC`/`DESC` ordering?

Reply:
AFAIK upper_bound will return the first position where the comparison function returns false (or the opposite, not sure). The comparison you need to use here is based solely on nullness. However, depending on whether nulls are at the beginning or end, you need to look for the first not-null, or first null, respectively. The pseudo code above should do this.
Does this answer your question? I may be missing some context.

Reply:
Hang on, `upper_bound()` searches a sorted list for a *value*, using the specified functor/lambda as the `operator < ()`. There isn't a "value" we're searching for, in this case. Please correct me if I'm mistaken.

I suspect we'd still need a (single!) `find_if()`, because we might not be able to `upper_bound()` on a partitioned column.

Reply:
P.S. Is there a version of `thrust::upper_bound()` that takes two lambdas? (One for `operator ==()`, another for `operator <()`.)

Reply:
Yeah, there's no value to search for. But if a function returns false for all elements left of an index, and true for all elements to the right of it, we can do a binary search with it (which is what upper_bound is. And in this case it is a lambda that checks nullness. AFAIK it should work, I can follow up with the team tomorrow and verify.

Reply:
Thank you for the advice and discussion, @vuule. From the reading, I think I agree that a dummy value should do, for the search key. I'll update here with the results of my tests.

Reply:
Based on suggestions from @vuule, @jrhemstad, @codereport, et al, I've switched over to using a `null_count()` based calculation in the non-grouped case, and `partition_point()` in the grouped one.

If it comes to it, I'll benchmark with the custom kernel that @harrism recommended. My guess at the moment is that null-timestamps aren't a major concern, save for safety.